### PR TITLE
Use fork in test test_cache_return_value_per_process

### PR DESCRIPTION
**Pull request by WeichenXu123**
Signed-off-by: Weichen Xu <weichen.xu@databricks.com>

## What changes are proposed in this pull request?

Use fork in test test_cache_return_value_per_process.
Reason:
Use fork to create child process instead of using `multiprocessing.Process` because
`multiprocessing.Process` creates child process by pickling the target function
and start a new process to run the pickled function. But the global variable
`_per_process_value_cache_map` dict content is not pickled, this make child process
automatically clear the `_per_process_value_cache_map` dict content.

## How is this patch tested?

Unit tests.

## Does this PR change the documentation?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly by following the steps below.

1. Check the status of the `ci/circleci: build_doc` check. If it's successful, proceed to the
   next step, otherwise fix it.
2. Click `Details` on the right to open the job page of CircleCI.
3. Click the `Artifacts` tab.
4. Click `docs/build/html/index.html`.
5. Find the changed pages / sections and make sure they render correctly.

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [x] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: tests/utils/test_process_utils.py [hunk 1]
```diff
@@ -54,10 +46,23 @@ def test_cache_return_value_per_process():
...
-    )
-    child_proc.start()
-    child_proc.join()
-    assert queue.get(), "Testing inside child process failed."
+    if os.name != "nt":
+        # Test child process invalidates the cache.
+        # We don't create child process by `multiprocessing.Process` because
+        # `multiprocessing.Process` creates child process by pickling the target function
+        # and start a new process to run the pickled function. But the global variable
+        # `_per_process_value_cache_map` dict content is not pickled, this make child process
```
Comment (harupy) [inline tests/utils/test_process_utils.py:54]:
```suggestion
        # `_per_process_value_cache_map` dict content is not pickled, this makes child process
```

File: tests/utils/test_process_utils.py [hunk 2]
```diff
@@ -54,10 +46,23 @@ def test_cache_return_value_per_process():
...
-        target=_test_cache_return_value_per_process_child_proc_target, args=(path1, path3, queue)
-    )
-    child_proc.start()
-    child_proc.join()
-    assert queue.get(), "Testing inside child process failed."
+    if os.name != "nt":
+        # Test child process invalidates the cache.
+        # We don't create child process by `multiprocessing.Process` because
+        # `multiprocessing.Process` creates child process by pickling the target function
+        # and start a new process to run the pickled function. But the global variable
```
Comment (harupy) [inline tests/utils/test_process_utils.py:53]:
```suggestion
        # and starts a new process to run the pickled function. But the global variable
```

File: tests/utils/test_process_utils.py [hunk 3]
```diff
@@ -54,10 +46,23 @@ def test_cache_return_value_per_process():
...
     assert len({path1, path3, f2_path1, f2_path2}) == 4
 
-    queue = Queue()
-    child_proc = Process(
-        target=_test_cache_return_value_per_process_child_proc_target, args=(path1, path3, queue)
-    )
-    child_proc.start()
-    child_proc.join()
-    assert queue.get(), "Testing inside child process failed."
+    if os.name != "nt":
```
Comment (harupy) [inline tests/utils/test_process_utils.py:49]: Why do we ignore windows?

File: tests/utils/test_process_utils.py [hunk 4]
```diff
@@ -54,10 +46,23 @@ def test_cache_return_value_per_process():
...
+        if pid > 0:
+            # in parent process
+            child_pid = pid
+            # check child process exit with return value 0.
+            assert os.waitpid(child_pid, 0)[1] == 0
+        else:
+            # in forked out child process
+            child_path1 = _gen_random_str1(True)
+            child_path2 = _gen_random_str1(False)
+            test_pass = len({path1, path3, child_path1, child_path2}) == 4
```
Comment (harupy) [inline tests/utils/test_process_utils.py:66]:
```suggestion
            assert len({path1, path3, child_path1, child_path2}) == 4
```

why not assert here?

File: tests/utils/test_process_utils.py [hunk 5]
```diff
@@ -54,10 +46,23 @@ def test_cache_return_value_per_process():
...
-    child_proc.start()
-    child_proc.join()
-    assert queue.get(), "Testing inside child process failed."
+    if os.name != "nt":
+        # Test child process invalidates the cache.
+        # We don't create child process by `multiprocessing.Process` because
+        # `multiprocessing.Process` creates child process by pickling the target function
+        # and start a new process to run the pickled function. But the global variable
+        # `_per_process_value_cache_map` dict content is not pickled, this make child process
+        # automatically clear the `_per_process_value_cache_map` dict content.
```
Comment (harupy) [inline tests/utils/test_process_utils.py:55]: `child process automatically clear the `_per_process_value_cache_map` dict content.` sounds a bit strange to me. I think I'd say `_per_process_value_cache_map is empty in the child process.`

File: tests/utils/test_process_utils.py [hunk 6]
```diff
@@ -54,10 +46,23 @@ def test_cache_return_value_per_process():
...
+            child_pid = pid
+            # check child process exit with return value 0.
+            assert os.waitpid(child_pid, 0)[1] == 0
+        else:
+            # in forked out child process
+            child_path1 = _gen_random_str1(True)
+            child_path2 = _gen_random_str1(False)
+            test_pass = len({path1, path3, child_path1, child_path2}) == 4
+            # exit forked out child process with exit code representing testing pass or fail.
+            os._exit(0 if test_pass else 1)
```
Comment (harupy) [inline tests/utils/test_process_utils.py:68]: Does this line terminate test execution?

File: tests/utils/test_process_utils.py (hunk 5 above)
Comment (harupy) [reply tests/utils/test_process_utils.py:55]: Or child and parent don't share the same global variable.

File: tests/utils/test_process_utils.py (hunk 6 above)
Comment (harupy) [reply tests/utils/test_process_utils.py:68]: Looks like it doesn't.

File: tests/utils/test_process_utils.py (hunk 4 above)
Comment (WeichenXu123) [reply tests/utils/test_process_utils.py:66]: `os._exit` makes forked out child process exit immediately, in contrast, if use `assert`, no matter assert pass or raise Assertion error, the pytest testing continues running until all tests completed. But my intention is to make child process exits immediately, the remaining tests it should be run only in parent process.

Comment (harupy) [reply tests/utils/test_process_utils.py:66]: I'm running the code below with pytest:

```python
import os


def test_foo():
    os.fork()
    os._exit(0)


def test_bar():
    os.fork()
    os._exit(1)

```

Comment (WeichenXu123) [reply tests/utils/test_process_utils.py:66]:
```
def test_foo():
    os.fork()
    os._exit(0)
```
your code call "os._exit" in both parent and child processes.

My code is:
```
if os.fork() > 0:
  # in parent process
  pass
else:
  # in child process
  os._exit(0)
```
```

Comment (WeichenXu123) [reply tests/utils/test_process_utils.py:67]: `os.fork` has 2 return point:
* in child process, it return 0,
* in parent process, it return the pid of child process (which is >0)

Comment (harupy) [reply tests/utils/test_process_utils.py:67]: Offline dicussed with Weichen and understood what's happening.

File: tests/utils/test_process_utils.py [hunk 7]
```diff
@@ -54,10 +46,24 @@ def test_cache_return_value_per_process():
...
-    child_proc.join()
-    assert queue.get(), "Testing inside child process failed."
+    # Skip the following block for Windows which doesn't support `os.fork`
+    if os.name != "nt":
+        # Test child process invalidates the cache.
+        # We don't create child process by `multiprocessing.Process` because
+        # `multiprocessing.Process` creates child process by pickling the target function
+        # and starts a new process to run the pickled function. But the global variable
+        # `_per_process_value_cache_map` dict content is not pickled, this makes child
+        #         # and parent don't share the same global variables.
```
Comment (harupy) [inline tests/utils/test_process_utils.py:56]:
```suggestion
        # and parent don't share the same global variables.
```

File: tests/utils/test_process_utils.py [hunk 8]
```diff
@@ -54,10 +46,24 @@ def test_cache_return_value_per_process():
...
-    child_proc.join()
-    assert queue.get(), "Testing inside child process failed."
+    # Skip the following block on Windows which doesn't support `os.fork`
+    if os.name != "nt":
+        # Test child process invalidates the cache.
+        # We don't create child process by `multiprocessing.Process` because
+        # `multiprocessing.Process` creates child process by pickling the target function
+        # and starts a new process to run the pickled function. But the global variable
+        # `_per_process_value_cache_map` dict content is not pickled, this makes child
+        #         # and parent don't share the same global variables.
```
Comment (harupy) [inline tests/utils/test_process_utils.py:56]: Can we simply say

```
Use `os.fork()` to make parent and child processes share the same global variable `_per_process_value_cache_map`
```

?

**Review by harupy — APPROVED** [review]
LGTM

**Pull request merged**